### PR TITLE
Use dagger hilt

### PR DESCRIPTION
**Pull request by quangson91**

### CODE REVIEW

File: data/build.gradle [hunk 1]
```diff
@@ -78,4 +74,8 @@ dependencies {
     kapt "com.github.bumptech.glide:compiler:$glide_version"
 
     implementation 'org.jsoup:jsoup:1.13.1'
+
+    def hilt_version = "2.28-alpha"
+    implementation("com.google.dagger:hilt-android:$hilt_version")
+    kapt("com.google.dagger:hilt-android-compiler:$hilt_version")
```
Comment (jenthone) [inline data/build.gradle:80]: ( ) is unnecessary

File: data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt [hunk 2]
```diff
@@ -1,63 +1,48 @@
...
-            .addConverterFactory(MoshiConverterFactory.create(get()))
-            .baseUrl(Const.BASE_URL)
-            .client(get())
-            .build()
+import javax.inject.Singleton
+
+@Module
+@InstallIn(ApplicationComponent::class)
+object DataModule {
+
```
Comment (jenthone) [inline data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt:19]: Please remove this blank line

File: data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt [hunk 3]
```diff
@@ -1,63 +1,48 @@
...
+        @ApplicationContext context: Context
+    ): AppDatabase = AppDatabase.getInstance(context)
+
+    @Provides
+    @Singleton
+    fun provideItemDao(
+        appDatabase: AppDatabase
+    ): ItemDao = appDatabase.itemDao()
+
+
```
Comment (jenthone) [inline data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt:32]: Please remove one blank line

File: data/src/main/java/io/github/jenthone/hackernews/data/di/NetworkModule.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import okhttp3.OkHttpClient
+import retrofit2.Retrofit
+import retrofit2.converter.moshi.MoshiConverterFactory
+import java.util.concurrent.TimeUnit
+import javax.inject.Singleton
+
+@Module
+@InstallIn(ApplicationComponent::class)
+object NetworkModule {
+
```
Comment (jenthone) [inline data/src/main/java/io/github/jenthone/hackernews/data/di/NetworkModule.kt:21]: Ditto

File: data/src/main/java/io/github/jenthone/hackernews/data/di/RepositoryModule.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import dagger.hilt.android.scopes.ActivityRetainedScoped
+import io.github.jenthone.hackernews.data.repository.ItemRepositoryImpl
+import io.github.jenthone.hackernews.data.repository.StoryRepositoryImpl
+import io.github.jenthone.hackernews.domain.repository.ItemRepository
+import io.github.jenthone.hackernews.domain.repository.StoryRepository
+
+@Module
+@InstallIn(ActivityRetainedComponent::class)
+abstract class RepositoryModule {
+
```
Comment (jenthone) [inline data/src/main/java/io/github/jenthone/hackernews/data/di/RepositoryModule.kt:16]: Ditto

File: presentation/build.gradle [hunk 6]
```diff
@@ -44,22 +46,31 @@ dependencies {
...
+    implementation "androidx.fragment:fragment-ktx:1.2.5"
+
     implementation 'org.ocpsoft.prettytime:prettytime:4.0.4.Final'
+
+    def hilt_version = "2.28-alpha"
+    implementation("com.google.dagger:hilt-android:$hilt_version")
+    kapt("com.google.dagger:hilt-android-compiler:$hilt_version")
+    def hilt_viewmodel_version = "1.0.0-alpha01"
+    implementation("androidx.hilt:hilt-lifecycle-viewmodel:$hilt_viewmodel_version")
+    kapt("androidx.hilt:hilt-compiler:$hilt_viewmodel_version")
```
Comment (jenthone) [inline presentation/build.gradle:75]: Please remove ( ) wrapper

File: presentation/src/main/java/io/github/jenthone/hackernews/helper/CommonHelper.kt [hunk 7]
```diff
@@ -3,14 +3,18 @@ package io.github.jenthone.hackernews.helper
...
 }
 
-fun <T> LiveData<T>.observe(owner: LifecycleOwner, action: (T) -> Unit) {
-    observe(owner, Observer { action(it) })
+fun <T> LiveData<T>.observeNonNull(owner: LifecycleOwner, action: (T) -> Unit) {
+    observe(owner, Observer {
+        if (it != null) {
+            action(it)
+        }
+    })
```
Comment (jenthone) [inline presentation/src/main/java/io/github/jenthone/hackernews/helper/CommonHelper.kt:19]: We can use
```kotlin
fun <T> LiveData<T>.observeNonNull(owner: LifecycleOwner, action: (T) -> Unit) =
    observe(owner, Observer { it?.let(action) })
```

Comment (jenthone) [reply presentation/src/main/java/io/github/jenthone/hackernews/helper/CommonHelper.kt:19]: And should we change function name to `observeNotNull`

File: data/build.gradle (hunk 1 above)
Comment (quangson91) [reply data/build.gradle:80]: Done.

File: data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt (hunk 2 above)
Comment (quangson91) [reply data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt:19]: Done.

File: data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt (hunk 3 above)
Comment (quangson91) [reply data/src/main/java/io/github/jenthone/hackernews/data/di/DataModule.kt:32]: Done.

File: presentation/src/main/java/io/github/jenthone/hackernews/helper/CommonHelper.kt (hunk 7 above)
Comment (quangson91) [reply presentation/src/main/java/io/github/jenthone/hackernews/helper/CommonHelper.kt]: Done

File: presentation/build.gradle (hunk 6 above)
Comment (quangson91) [reply presentation/build.gradle]: Done

File: data/src/main/java/io/github/jenthone/hackernews/data/di/RepositoryModule.kt (hunk 5 above)
Comment (quangson91) [reply data/src/main/java/io/github/jenthone/hackernews/data/di/RepositoryModule.kt]: Done

File: data/src/main/java/io/github/jenthone/hackernews/data/di/NetworkModule.kt (hunk 4 above)
Comment (quangson91) [reply data/src/main/java/io/github/jenthone/hackernews/data/di/NetworkModule.kt]: Done

File: build.gradle [hunk 8]
```diff
@@ -15,8 +15,9 @@ buildscript {
         
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.6.1'
+        classpath 'com.android.tools.build:gradle:4.0.1'
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+        classpath("com.google.dagger:hilt-android-gradle-plugin:2.28-alpha")
```
Comment (jenthone) [inline build.gradle:20]: Sorry, we should also remove ( )

**Pull request merged**